File: services/inventory-service/internal/store/store.go
// Package store is the only place in this service that knows SQL.
//
// Every rule that must hold under concurrency is enforced by Postgres — a row
// lock, a unique index or a CHECK — not by Go. Two pods, a retry storm and a
// reaper all run against the same tables, and application-level "check then
// write" loses that race every time.
package store

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/sayanc422/aquahub/services/inventory-service/internal/inventory"
)

var (
	// ErrNotFound — no reservation with that id.
	ErrNotFound = errors.New("reservation not found")
	// ErrIdempotencyConflict — the key has been used for a different request.
	ErrIdempotencyConflict = errors.New("idempotency key reused with a different request")
	// ErrStateConflict — the reservation cannot make this transition.
	ErrStateConflict = errors.New("reservation is not in a state that allows this")
	// ErrExpired — the hold ran out before the caller committed it.
	ErrExpired = errors.New("reservation expired")
)

// States of a reservation. A hold is the only state that subtracts from
// availability, and it is the only state with a deadline.
const (
	StateHeld      = "held"
	StateCommitted = "committed"
	StateReleased  = "released"
	StateExpired   = "expired"
)

type Store struct{ pool *pgxpool.Pool }

func New(pool *pgxpool.Pool) *Store { return &Store{pool: pool} }

// Line is one tank's contribution to a reservation.
type Line struct {
	TankID   int64  `json:"-"`
	TankCode string `json:"tank"`
	Quantity int    `json:"quantity"`
}

// Reservation is the stored form. The API shape lives in the api package; this
// is the service's own record.
type Reservation struct {
	ID        uuid.UUID
	OrderRef  string
	SKU       string
	Quantity  int
	State     string
	ExpiresAt time.Time
	CreatedAt time.Time
	Lines     []Line
}

// NewRequest is one attempt to hold stock.
type NewRequest struct {
	IdempotencyKey string
	RequestDigest  string
	OrderRef       string
	SKU            string
	Quantity       int
	TTL            time.Duration
}

// TankStock is the per-tank view returned by the stock endpoint.
type TankStock struct {
	Code      string `json:"tank"`
	Status    string `json:"status"`
	OnHand    int    `json:"onHand"`
	Held      int    `json:"held"`
	Available int    `json:"available"`
	Note      string `json:"note,omitempty"`
}

// Availability is read through this expression everywhere, and `now()` is the
// reason the service is correct without the reaper: an expired hold stops
// subtracting the instant it expires, not when a background job notices. The
// reaper only tidies the state column so the table reads honestly.
const availableExpr = `
	t.quantity_on_hand - COALESCE((
		SELECT SUM(l.quantity)
		  FROM reservation_line l
		  JOIN reservation r ON r.id = l.reservation_id
		 WHERE l.tank_id = t.id
		   AND r.state = 'held'
		   AND r.expires_at > now()
	), 0)`

// Reserve creates a hold, or replays the one this idempotency key already made.
//
// The second return value is true when the call was a replay: the caller turns
// that into 200 rather than 201, so a client that retries through a lost
// response can tell a fresh hold from an echo of its own.
func (s *Store) Reserve(ctx context.Context, req NewRequest) (*Reservation, bool, error) {
	var (
		res      *Reservation
		replayed bool
	)
	err := pgx.BeginFunc(ctx, s.pool, func(tx pgx.Tx) error {
		id := uuid.New()
		var (
			existingID     uuid.UUID
			existingDigest string
			inserted       bool
		)
		// ON CONFLICT DO UPDATE, not DO NOTHING. DO NOTHING returns no row when
		// the conflicting insert is still uncommitted in another transaction,
		// and the follow-up SELECT cannot see it either — the retry would look
		// like a lost key. DO UPDATE takes the row lock and waits for that
		// transaction to finish, so the loser of the race always ends up
		// holding the winner's row. `xmax = 0` distinguishes the two: on a
		// genuine insert the row has no updating transaction id.
		err := tx.QueryRow(ctx, `
			INSERT INTO reservation
			    (id, idempotency_key, request_digest, order_ref, sku, quantity, state, expires_at)
			VALUES ($1, $2, $3, $4, $5, $6, 'held', now() + $7::interval)
			ON CONFLICT (idempotency_key)
			    DO UPDATE SET updated_at = reservation.updated_at
			RETURNING id, request_digest, (xmax = 0) AS inserted`,
			id, req.IdempotencyKey, req.RequestDigest, req.OrderRef, req.SKU, req.Quantity,
			fmt.Sprintf("%d seconds", int(req.TTL.Seconds())),
		).Scan(&existingID, &existingDigest, &inserted)
		if err != nil {
			return fmt.Errorf("insert reservation: %w", err)
		}

		if !inserted {
			if existingDigest != req.RequestDigest {
				return ErrIdempotencyConflict
			}
			loaded, err := loadReservation(ctx, tx, existingID)
			if err != nil {
				return err
			}
			res, replayed = loaded, true
			return nil
		}

		// Locking and reading availability are two statements, deliberately.
		//
		// Under READ COMMITTED a statement's snapshot is taken when the statement
		// begins — before it blocks on a row lock. Doing both in one
		// `SELECT ... FOR UPDATE` therefore returns availability as it looked
		// *before* the transaction ahead of it inserted its lines, and two
		// reservations promise the same fish. The test that proves it is
		// TestConcurrentReservationsCannotOversell, and it failed against exactly
		// that one-statement version.
		//
		// So: take the locks first, then read in a second statement, whose fresh
		// snapshot includes everything the transaction we waited for committed.
		//
		// `ORDER BY id` makes the lock order total, so two reservations that
		// overlap on two tanks cannot deadlock. The tank row is the concurrency
		// unit even though this transaction never updates it: it is the row that
		// names the resource being promised.
		var tankIDs []int64
		lockRows, err := tx.Query(ctx, `
			SELECT id FROM tank
			 WHERE sku = $1 AND status = 'open'
			 ORDER BY id
			   FOR UPDATE`, req.SKU)
		if err != nil {
			return fmt.Errorf("lock tanks: %w", err)
		}
		for lockRows.Next() {
			var id int64
			if err := lockRows.Scan(&id); err != nil {
				lockRows.Close()
				return err
			}
			tankIDs = append(tankIDs, id)
		}
		lockRows.Close()
		if err := lockRows.Err(); err != nil {
			return err
		}
		if len(tankIDs) == 0 {
			return inventory.ErrInsufficientStock
		}

		// Only the tanks actually locked above. A tank added for this SKU between
		// the two statements is not covered by our locks, so allocating from it
		// would reopen the same race on a new row.
		rows, err := tx.Query(ctx, `
			SELECT t.id, t.code, `+availableExpr+`
			  FROM tank t
			 WHERE t.id = ANY($1)
			 ORDER BY t.id`, tankIDs)
		if err != nil {
			return fmt.Errorf("read availability: %w", err)
		}
		var avail []inventory.TankAvailability
		for rows.Next() {
			var a inventory.TankAvailability
			if err := rows.Scan(&a.TankID, &a.Code, &a.Available); err != nil {
				rows.Close()
				return err
			}
			avail = append(avail, a)
		}
		rows.Close()
		if err := rows.Err(); err != nil {
			return err
		}

		plan, err := inventory.Plan(req.Quantity, avail)
		if err != nil {
			// Rolls back the reservation row with it. A refused reservation
			// must not burn the idempotency key: the customer adds one fewer
			// fish and retries with the same key, and that has to work.
			return err
		}

		batch := &pgx.Batch{}
		for _, line := range plan {
			batch.Queue(`INSERT INTO reservation_line (reservation_id, tank_id, quantity)
			             VALUES ($1, $2, $3)`, id, line.TankID, line.Quantity)
		}
		if err := tx.SendBatch(ctx, batch).Close(); err != nil {
			return fmt.Errorf("insert reservation lines: %w", err)
		}

		loaded, err := loadReservation(ctx, tx, id)
		if err != nil {
			return err
		}
		res = loaded
		return nil
	})
	if err != nil {
		return nil, false, err
	}
	return res, replayed, nil
}

// Commit turns a hold into a sale: the stock leaves the tank for good.
//
// Idempotent — committing an already-committed reservation returns it
// unchanged, because the order service will retry this call after a timeout and
// must not be told its own completed work failed.
func (s *Store) Commit(ctx context.Context, id uuid.UUID) (*Reservation, error) {
	var res *Reservation
	err := pgx.BeginFunc(ctx, s.pool, func(tx pgx.Tx) error {
		var state string
		var expiresAt time.Time
		err := tx.QueryRow(ctx,
			`SELECT state, expires_at FROM reservation WHERE id = $1 FOR UPDATE`, id).
			Scan(&state, &expiresAt)
		if errors.Is(err, pgx.ErrNoRows) {
			return ErrNotFound
		}
		if err != nil {
			return err
		}

		switch state {
		case StateCommitted:
			loaded, err := loadReservation(ctx, tx, id)
			res = loaded
			return err
		case StateReleased:
			return ErrStateConflict
		case StateExpired:
			return ErrExpired
		}

		// A hold whose deadline has passed is expired whether or not the reaper
		// has been round yet. Committing it would sell stock the availability
		// query has already promised to somebody else.
		if !expiresAt.After(time.Now()) {
			if _, err := tx.Exec(ctx,
				`UPDATE reservation SET state = 'expired', updated_at = now() WHERE id = $1`, id); err != nil {
				return err
			}
			return ErrExpired
		}

		// Locking order matches Reserve: tanks ascending by id.
		if _, err := tx.Exec(ctx, `
			UPDATE tank t
			   SET quantity_on_hand = t.quantity_on_hand - l.quantity
			  FROM reservation_line l
			 WHERE l.reservation_id = $1 AND t.id = l.tank_id`, id); err != nil {
			// quantity_on_hand >= 0 is a CHECK. If this ever fires, the bug is
			// in this service, and a violated constraint is the correct outcome:
			// a refused commit beats a tank holding minus four fish.
			var pgErr *pgconn.PgError
			if errors.As(err, &pgErr) && pgErr.Code == "23514" {
				return fmt.Errorf("commit would drive stock negative: %w", err)
			}
			return err
		}

		if _, err := tx.Exec(ctx,
			`UPDATE reservation SET state = 'committed', updated_at = now() WHERE id = $1`, id); err != nil {
			return err
		}
		loaded, err := loadReservation(ctx, tx, id)
		res = loaded
		return err
	})
	if err != nil {
		return nil, err
	}
	return res, nil
}

// Release returns a hold to stock early — an abandoned cart, a failed payment.
// Releasing an already-released or already-expired hold is a no-op, so the
// caller's retry and the reaper cannot fight over it.
func (s *Store) Release(ctx context.Context, id uuid.UUID) (*Reservation, error) {
	var res *Reservation
	err := pgx.BeginFunc(ctx, s.pool, func(tx pgx.Tx) error {
		var state string
		err := tx.QueryRow(ctx, `SELECT state FROM reservation WHERE id = $1 FOR UPDATE`, id).Scan(&state)
		if errors.Is(err, pgx.ErrNoRows) {
			return ErrNotFound
		}
		if err != nil {
			return err
		}
		if state == StateCommitted {
			// Undoing a sale is a refund, and refunds belong to
			// payment-service. This service will not silently reverse one.
			return ErrStateConflict
		}
		if state == StateHeld {
			if _, err := tx.Exec(ctx,
				`UPDATE reservation SET state = 'released', updated_at = now() WHERE id = $1`, id); err != nil {
				return err
			}
		}
		loaded, err := loadReservation(ctx, tx, id)
		res = loaded
		return err
	})
	if err != nil {
		return nil, err
	}
	return res, nil
}

// Get returns one reservation.
func (s *Store) Get(ctx context.Context, id uuid.UUID) (*Reservation, error) {
	return loadReservation(ctx, s.pool, id)
}

// ExpireDue marks every hold past its deadline as expired and returns how many.
//
// This changes no availability — the availability query already ignores them.
// It exists so the table says what is true, so the active-holds gauge is not
// permanently wrong, and so a support question ("what happened to this hold?")
// has an answer other than arithmetic on a timestamp.
func (s *Store) ExpireDue(ctx context.Context) (int64, error) {
	tag, err := s.pool.Exec(ctx, `
		UPDATE reservation
		   SET state = 'expired', updated_at = now()
		 WHERE state = 'held' AND expires_at <= now()`)
	if err != nil {
		return 0, err
	}
	return tag.RowsAffected(), nil
}

// ActiveHolds counts holds that still subtract from availability.
func (s *Store) ActiveHolds(ctx context.Context) (int64, error) {
	var n int64
	err := s.pool.QueryRow(ctx,
		`SELECT count(*) FROM reservation WHERE state = 'held' AND expires_at > now()`).Scan(&n)
	return n, err
}

// Stock reports every tank holding a SKU, quarantined tanks included.
//
// A quarantined tank shows its count and contributes nothing to availability.
// Hiding it would make the shop floor and the API disagree, and the person
// reconciling them is standing in front of the tank.
func (s *Store) Stock(ctx context.Context, sku string) ([]TankStock, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT t.code, t.status, t.quantity_on_hand, COALESCE(t.note, ''),
		       COALESCE((
		           SELECT SUM(l.quantity)
		             FROM reservation_line l
		             JOIN reservation r ON r.id = l.reservation_id
		            WHERE l.tank_id = t.id AND r.state = 'held' AND r.expires_at > now()
		       ), 0) AS held
		  FROM tank t
		 WHERE t.sku = $1
		 ORDER BY t.code`, sku)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := []TankStock{}
	for rows.Next() {
		var ts TankStock
		if err := rows.Scan(&ts.Code, &ts.Status, &ts.OnHand, &ts.Note, &ts.Held); err != nil {
			return nil, err
		}
		if ts.Status == "open" {
			ts.Available = ts.OnHand - ts.Held
		}
		out = append(out, ts)
	}
	return out, rows.Err()
}

// Ping is the readiness check: a pool that cannot reach Postgres must leave the
// Service endpoints rather than serve 500s.
func (s *Store) Ping(ctx context.Context) error { return s.pool.Ping(ctx) }

// querier is satisfied by both *pgxpool.Pool and pgx.Tx, so reservations load
// the same way inside and outside a transaction.
type querier interface {
	QueryRow(context.Context, string, ...any) pgx.Row
	Query(context.Context, string, ...any) (pgx.Rows, error)
}

func loadReservation(ctx context.Context, q querier, id uuid.UUID) (*Reservation, error) {
	var r Reservation
	err := q.QueryRow(ctx, `
		SELECT id, order_ref, sku, quantity, state, expires_at, created_at
		  FROM reservation WHERE id = $1`, id).
		Scan(&r.ID, &r.OrderRef, &r.SKU, &r.Quantity, &r.State, &r.ExpiresAt, &r.CreatedAt)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, ErrNotFound
	}
	if err != nil {
		return nil, err
	}

	// A hold past its deadline reads as expired even before the reaper has
	// updated the row, so the API never reports a hold the availability query
	// has already stopped honouring.
	if r.State == StateHeld && !r.ExpiresAt.After(time.Now()) {
		r.State = StateExpired
	}

	rows, err := q.Query(ctx, `
		SELECT t.code, l.tank_id, l.quantity
		  FROM reservation_line l JOIN tank t ON t.id = l.tank_id
		 WHERE l.reservation_id = $1
		 ORDER BY t.code`, id)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	r.Lines = []Line{}
	for rows.Next() {
		var l Line
		if err := rows.Scan(&l.TankCode, &l.TankID, &l.Quantity); err != nil {
			return nil, err
		}
		r.Lines = append(r.Lines, l)
	}
	return &r, rows.Err()
}

File: services/payment-service/src/store.rs
//! Persistence. The only module that writes SQL.
//!
//! Two rules run through all of it:
//!
//!   * **The intent is written before the acquirer is called.** A service that
//!     charges first and records afterwards loses the record of every charge it
//!     dies in the middle of, and there is then nothing to reconcile against.
//!   * **Every state change is one transaction that writes both the payment row
//!     and its ledger entry.** A balance that disagrees with the entries behind
//!     it is unauditable, and the entries are the source of truth.

use crate::ledger::{Payment, PaymentState, Transition};
use crate::money::{Currency, Money};
use chrono::{DateTime, Utc};
use sqlx::{postgres::PgPool, Row};
use uuid::Uuid;

/// The migrations, embedded at compile time so no database is needed to build.
/// Exposed here rather than run from `main` so the tests apply exactly the same
/// schema the service does — a test suite against a hand-written schema proves
/// the hand-written schema.
pub static MIGRATOR: sqlx::migrate::Migrator = sqlx::migrate!("./migrations");

#[derive(Debug, thiserror::Error)]
pub enum StoreError {
    #[error("database: {0}")]
    Db(#[from] sqlx::Error),
    #[error("payment not found")]
    NotFound,
    #[error("stored payment {0} is not readable: {1}")]
    Corrupt(Uuid, String),
    #[error("idempotency key reused for a different request")]
    IdempotencyConflict,
}

/// A payment as stored, with the identity the machine does not carry.
#[derive(Debug, Clone)]
pub struct StoredPayment {
    pub id: Uuid,
    pub idempotency_key: String,
    pub order_reference: String,
    pub payment: Payment,
    pub acquirer_ref: Option<String>,
    pub created_at: DateTime<Utc>,
}

#[derive(Debug, Clone)]
pub struct LedgerRow {
    pub seq: i32,
    pub kind: String,
    pub amount_minor: i64,
    pub balance_after: i64,
    pub detail: Option<String>,
    pub at: DateTime<Utc>,
}

#[derive(Clone)]
pub struct Store {
    pool: PgPool,
}

impl Store {
    pub fn new(pool: PgPool) -> Self {
        Store { pool }
    }

    pub async fn ping(&self) -> Result<(), StoreError> {
        sqlx::query("SELECT 1").execute(&self.pool).await?;
        Ok(())
    }

    /// Write the intent, or hand back the payment this key already made.
    ///
    /// `ON CONFLICT DO UPDATE`, not `DO NOTHING`, for the same reason as in
    /// inventory-service: `DO NOTHING` returns no row when the conflicting
    /// insert is still uncommitted elsewhere, and the follow-up `SELECT` cannot
    /// see it either, so a concurrent retry would look like a lost key.
    /// `DO UPDATE` takes the row lock and waits. `xmax = 0` separates a genuine
    /// insert from a conflict.
    ///
    /// Returns `(payment, true)` when this call created it.
    pub async fn begin_payment(
        &self,
        idempotency_key: &str,
        order_reference: &str,
        amount: Money,
    ) -> Result<(StoredPayment, bool), StoreError> {
        let mut tx = self.pool.begin().await?;

        let row = sqlx::query(
            r#"
            INSERT INTO payment (id, idempotency_key, order_reference, amount_minor, currency, state)
            VALUES ($1, $2, $3, $4, $5, 'pending')
            ON CONFLICT (idempotency_key)
                DO UPDATE SET updated_at = payment.updated_at
            RETURNING id, idempotency_key, order_reference, amount_minor, currency,
                      state, balance_minor, acquirer_ref, created_at, (xmax = 0) AS inserted
            "#,
        )
        .bind(Uuid::new_v4())
        .bind(idempotency_key)
        .bind(order_reference)
        .bind(amount.minor())
        .bind(amount.currency().code())
        .fetch_one(&mut *tx)
        .await?;

        let inserted: bool = row.try_get("inserted")?;
        let stored = stored_from_row(&row)?;

        if !inserted {
            // Same key, different money, is a caller bug worth refusing rather
            // than quietly answering about a different payment.
            if stored.payment.amount != amount || stored.order_reference != order_reference {
                return Err(StoreError::IdempotencyConflict);
            }
            tx.commit().await?;
            return Ok((stored, false));
        }

        // Entry 1 is the intent itself, written before the acquirer is called.
        sqlx::query(
            r#"
            INSERT INTO ledger_entry (payment_id, seq, kind, amount_minor, balance_after, detail)
            VALUES ($1, 1, 'authorise', $2, 0, $3)
            "#,
        )
        .bind(stored.id)
        .bind(amount.minor())
        .bind(format!("intent for order {order_reference}"))
        .execute(&mut *tx)
        .await?;

        tx.commit().await?;
        Ok((stored, true))
    }

    /// Record the outcome of a transition: the payment row and its entry, in
    /// one transaction.
    ///
    /// The `WHERE state = $current` clause is the concurrency control. Two
    /// resolutions of the same pending payment — an inline one and the
    /// reconciler — can race, and the loser writes nothing rather than
    /// appending a second capture for the same charge.
    pub async fn apply(
        &self,
        id: Uuid,
        from: PaymentState,
        transition: Transition,
        acquirer_ref: Option<&str>,
        detail: &str,
    ) -> Result<bool, StoreError> {
        let mut tx = self.pool.begin().await?;

        let updated = sqlx::query(
            r#"
            UPDATE payment
               SET state = $1, balance_minor = $2, updated_at = now(),
                   acquirer_ref = COALESCE($3, acquirer_ref)
             WHERE id = $4 AND state = $5
            "#,
        )
        .bind(transition.state.as_str())
        .bind(transition.balance.minor())
        .bind(acquirer_ref)
        .bind(id)
        .bind(from.as_str())
        .execute(&mut *tx)
        .await?;

        if updated.rows_affected() == 0 {
            // Somebody else resolved it first. Not an error: the payment has an
            // outcome, which is all the caller needed.
            tx.rollback().await?;
            return Ok(false);
        }

        sqlx::query(
            r#"
            INSERT INTO ledger_entry (payment_id, seq, kind, amount_minor, balance_after, detail)
            SELECT $1, COALESCE(MAX(seq), 0) + 1, $2, $3, $4, $5
              FROM ledger_entry WHERE payment_id = $1
            "#,
        )
        .bind(id)
        .bind(transition.entry.as_str())
        .bind(transition.entry_amount.minor())
        .bind(transition.balance.minor())
        .bind(detail)
        .execute(&mut *tx)
        .await?;

        tx.commit().await?;
        Ok(true)
    }

    /// Apply a refund and remember the key that asked for it, in one
    /// transaction. A retried refund finds the key and pays out nothing more.
    pub async fn apply_refund(
        &self,
        id: Uuid,
        from: PaymentState,
        transition: Transition,
        refund_key: &str,
        detail: &str,
    ) -> Result<bool, StoreError> {
        let mut tx = self.pool.begin().await?;

        let updated = sqlx::query(
            r#"
            UPDATE payment SET state = $1, balance_minor = $2, updated_at = now()
             WHERE id = $3 AND state = $4
            "#,
        )
        .bind(transition.state.as_str())
        .bind(transition.balance.minor())
        .bind(id)
        .bind(from.as_str())
        .execute(&mut *tx)
        .await?;

        if updated.rows_affected() == 0 {
            tx.rollback().await?;
            return Ok(false);
        }

        let seq: i32 = sqlx::query(
            r#"
            INSERT INTO ledger_entry (payment_id, seq, kind, amount_minor, balance_after, detail)
            SELECT $1, COALESCE(MAX(seq), 0) + 1, 'refund', $2, $3, $4
              FROM ledger_entry WHERE payment_id = $1
            RETURNING seq
            "#,
        )
        .bind(id)
        .bind(transition.entry_amount.minor())
        .bind(transition.balance.minor())
        .bind(detail)
        .fetch_one(&mut *tx)
        .await?
        .try_get("seq")?;

        sqlx::query(
            r#"
            INSERT INTO refund_request (idempotency_key, payment_id, amount_minor, seq)
            VALUES ($1, $2, $3, $4)
            "#,
        )
        .bind(refund_key)
        .bind(id)
        .bind(transition.entry_amount.minor())
        .bind(seq)
        .execute(&mut *tx)
        .await?;

        tx.commit().await?;
        Ok(true)
    }

    pub async fn find_refund(&self, refund_key: &str) -> Result<Option<Uuid>, StoreError> {
        let row = sqlx::query("SELECT payment_id FROM refund_request WHERE idempotency_key = $1")
            .bind(refund_key)
            .fetch_optional(&self.pool)
            .await?;
        Ok(row.map(|r| r.get("payment_id")))
    }

    pub async fn get(&self, id: Uuid) -> Result<StoredPayment, StoreError> {
        let row = sqlx::query(
            r#"SELECT id, idempotency_key, order_reference, amount_minor, currency,
                      state, balance_minor, acquirer_ref, created_at
                 FROM payment WHERE id = $1"#,
        )
        .bind(id)
        .fetch_optional(&self.pool)
        .await?
        .ok_or(StoreError::NotFound)?;
        stored_from_row(&row)
    }

    pub async fn get_by_key(&self, key: &str) -> Result<StoredPayment, StoreError> {
        let row = sqlx::query(
            r#"SELECT id, idempotency_key, order_reference, amount_minor, currency,
                      state, balance_minor, acquirer_ref, created_at
                 FROM payment WHERE idempotency_key = $1"#,
        )
        .bind(key)
        .fetch_optional(&self.pool)
        .await?
        .ok_or(StoreError::NotFound)?;
        stored_from_row(&row)
    }

    pub async fn ledger(&self, id: Uuid) -> Result<Vec<LedgerRow>, StoreError> {
        let rows = sqlx::query(
            r#"SELECT seq, kind, amount_minor, balance_after, detail, at
                 FROM ledger_entry WHERE payment_id = $1 ORDER BY seq"#,
        )
        .bind(id)
        .fetch_all(&self.pool)
        .await?;

        Ok(rows
            .iter()
            .map(|r| LedgerRow {
                seq: r.get("seq"),
                kind: r.get("kind"),
                amount_minor: r.get("amount_minor"),
                balance_after: r.get("balance_after"),
                detail: r.get("detail"),
                at: r.get("at"),
            })
            .collect())
    }

    /// Payments the acquirer never answered for, old enough to be worth asking
    /// about again.
    pub async fn pending_since(
        &self,
        seconds: i64,
        limit: i64,
    ) -> Result<Vec<StoredPayment>, StoreError> {
        let rows = sqlx::query(
            r#"SELECT id, idempotency_key, order_reference, amount_minor, currency,
                      state, balance_minor, acquirer_ref, created_at
                 FROM payment
                WHERE state = 'pending' AND created_at < now() - make_interval(secs => $1)
                ORDER BY created_at
                LIMIT $2"#,
        )
        .bind(seconds as f64)
        .bind(limit)
        .fetch_all(&self.pool)
        .await?;

        rows.iter().map(stored_from_row).collect()
    }

    pub async fn count_pending(&self) -> Result<i64, StoreError> {
        let row = sqlx::query("SELECT count(*) AS n FROM payment WHERE state = 'pending'")
            .fetch_one(&self.pool)
            .await?;
        Ok(row.get("n"))
    }
}

fn stored_from_row(row: &sqlx::postgres::PgRow) -> Result<StoredPayment, StoreError> {
    let id: Uuid = row.try_get("id")?;
    let currency_code: String = row.try_get("currency")?;
    let currency =
        Currency::parse(&currency_code).map_err(|e| StoreError::Corrupt(id, e.to_string()))?;
    let amount = Money::new(row.try_get::<i64, _>("amount_minor")?, currency)
        .map_err(|e| StoreError::Corrupt(id, e.to_string()))?;
    let balance = Money::new(row.try_get::<i64, _>("balance_minor")?, currency)
        .map_err(|e| StoreError::Corrupt(id, e.to_string()))?;
    let state_text: String = row.try_get("state")?;
    let state = PaymentState::parse(&state_text)
        .ok_or_else(|| StoreError::Corrupt(id, format!("unknown state {state_text}")))?;

    Ok(StoredPayment {
        id,
        idempotency_key: row.try_get("idempotency_key")?,
        order_reference: row.try_get("order_reference")?,
        payment: Payment {
            state,
            amount,
            balance,
        },
        acquirer_ref: row.try_get("acquirer_ref")?,
        created_at: row.try_get("created_at")?,
    })
}
